### PR TITLE
#2548 #1439 support for `exemplify()` and `.examples`

### PR DESCRIPTION
**Pull request by StephanMeijer**
See:
- #2548 
- #1439

### CODE REVIEW

**Comment by cupcakearmy** [comment]
This would be an awesome ground work for https://github.com/StefanTerdell/zod-to-json-schema

**Comment by thelinuxlich** [comment]
+1

File: README.md [hunk 1]
```diff
@@ -2293,6 +2294,17 @@ documentedString.description; // A useful bit of text…
 
 This can be useful for documenting a field, for example in a JSON Schema using a library like [`zod-to-json-schema`](https://github.com/StefanTerdell/zod-to-json-schema)).
 
+### `.exemplify`
+
+Use `.exemplify()` to add to the `examples` property to the resulting schema.
+
+```ts
+const exemplifiedString = z.string().example("Lorem ipsum dolor sit amet.");
```
Comment (stephenwf) [inline README.md:2302]:
```suggestion
const exemplifiedString = z.string().exemplify("Lorem ipsum dolor sit amet.");
```

File: deno/lib/README.md [hunk 2]
```diff
@@ -2293,6 +2294,17 @@ documentedString.description; // A useful bit of text…
 
 This can be useful for documenting a field, for example in a JSON Schema using a library like [`zod-to-json-schema`](https://github.com/StefanTerdell/zod-to-json-schema)).
 
+### `.exemplify`
+
+Use `.exemplify()` to add to the `examples` property to the resulting schema.
+
+```ts
+const exemplifiedString = z.string().example("Lorem ipsum dolor sit amet.");
```
Comment (stephenwf) [inline deno/lib/README.md:2302]:
```suggestion
const exemplifiedString = z.string().exemplify("Lorem ipsum dolor sit amet.");
```

File: deno/lib/types.ts [hunk 3]
```diff
@@ -56,6 +56,7 @@ export type CustomErrorParams = Partial<util.Omit<ZodCustomIssue, "code">>;
 export interface ZodTypeDef {
   errorMap?: ZodErrorMap;
   description?: string;
+  examples?: unknown[];
```
Comment (stephenwf) [inline deno/lib/types.ts:59]: Could these use the `_output` to get types?

Possibly:
```ts
ParseReturnType<this["_output"]>
```

File: deno/lib/types.ts [hunk 4]
```diff
@@ -486,6 +500,19 @@ export abstract class ZodType<
     });
   }
 
+  exemplify(example: unknown): this {
```
Comment (stephenwf) [inline deno/lib/types.ts:503]: Having type safety for examples seems like quite a important feature.
```ts
exemplify(example: ParseReturnType<this["_output"]>): this {
```

Or something similar would ensure that required changes to examples are caught if the definition changes.

**Comment by SamSalvatico** [comment]
Tried to improve the code following @stephenwf suggestions here: https://github.com/StephanMeijer/zod/pull/1

**Comment by MikeCxx** [comment]
why does it merge to main???
anyone knows?

**Comment by colinhacks** [comment]
The idea of attaching metadata to schemas is complex (see #2496 and other discussions). This proposed API is too JSON-Schema-specific imo. Zod needs a more general-purpose solution that can accommodate a range of use cases. I'll be putting up a metadata proposal soon. Thanks for the great work here though, I'll ping you when the RFC is up.

**Comment by PieterT2000** [comment]
Any status update on this?

**Comment by punkpeye** [comment]
Oh man, I really needed this. @colinhacks would you reconsider this?

**Comment by tzezar** [comment]
it would be amazing to implement such feature

**Pull request closed without merging**